proposals/init.md

```diff
@@ -0,0 +1,490 @@
+attributes or modreqs when emitting during metadata. These have different 
+trade offs that need to be considered.
+
+Annotating a property set accessor with a modreq declaration means CLI compliant
+compilers will ignore the accessor unless it understands the modreq. That means
+only compilers aware of `init` will read the member. Compilers unaware of 
+`init` will ignore the `set` member and hence will not accidentally treat the
+property as read / write. 
+
+The downside of modreq is `init` becomes a part of the binary signature of 
+the `set` accessor. Adding or removing `init` will break binary compatbility 
+of the application.
+
+Using attributes to annotate the `set` accessor means that only compilers which
+understand the attribute will know to limit access to it. A compiler unaware 
+of `init` will see it as a simple read / write property and allow access.
+
+This would seemingly mean this decision is a choice between extra safety at 
+the expense of binary compatibility. Digging in a bit the extra safety is not
+exactly what it seems. It will not for instance protect against the following
+circumstances:
+
+1. Reflection over `public` members
+1. The use of `dynamic` 
+1. Compilers that don't recognize modreqs
+
+It should also be considered that when we complete the IL verification rules 
+for .NET 5, `init` will be one of those rules. that means extra enforcement 
+will be gained from simply verifying compilers emitting verifiable IL.
+
+The primary languages for .NET (C#, F# and VB) will all be updated to 
+recognize these `init` accessors. Hence the only realistic scenario here is 
+when a C# 9 compiler emits `init` properties and they are seen by a C# 8 
+compiler. That is the trade off to consider and weigh against binary 
+compatibility.
+
+Note: this discussion applies to the accessor members only, not to fields. There
+is no value to be gained by using a modreq on a field. The `init` feature for 
+fields is a relaxation of an existing rule. All existing compilers already 
+support `readonly` and hence an attribute serves fine as a way to alert them
+that write access can be extended in certain circumstances.
+
+**Resolution**
+The feature will use a modreq to encode the property `init` setter. The
+compelling factors were (in no particular order):
+
+1. Desire to discourage older compilers from violating `init` semantics
+1. Desire to make adding or removing `init` in a `virtual` declaratio or 
+`interface` both a source and binary breaking change.
+
+Given there was also no significant support for removing `init` to be a 
+binary compatible change it made the choice of using modreq straight forward.
+
+### init vs. initonly
+Syntax debate time.
+
+**Resolution**
+There was no other syntax which was overwelming favored in LDM. For the time 
+being we will be moving forward with the initial design laid out in this 
+document. This may change as the feature progresses and LDM members have time
+to digest the design here.
```

Comment:
Consider talking about `init set` vs `init`, and us being open to making `init` a shorthand for `init set` in the future if we decide to make `init` a modifier for functions.